core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java

```diff
@@ -105,14 +100,13 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                     switch (currentFieldName) {
                     case "must":
                         query = parseContext.parseInnerQueryBuilder();
-                        if (query != null) {
-                            mustClauses.add(query);
-                        }
+                        mustClauses.add(query);
                         break;
                     case "should":
                         query = parseContext.parseInnerQueryBuilder();
-                        if (query != null) {
-                            shouldClauses.add(query);
+                        shouldClauses.add(query);
+                        // EmptyQueryBuilder does not add lucene query later, skip setting minuminShouldMatch
+                        if (query != EmptyQueryBuilder.PROTOTYPE) {
```

Comment:
We are still adding the EmptyQueryBuilder here in any case (before the `if`). The issue here is to set the minimumShouldMatch to at least `1`, if there is no other value supplied in the JSON _and_ we are parsing this inside another filter. This will have to happen only once for the first "real" (meaning non-null) inner query, so we can safely skip EmptyQuery here. The only thing that might go wrong here is that we set `minimumShouldMatch` to eagerly  for query != EmptyQueryBuilder but that for some reason will later be resolved to `null` in `toQuery()`. I'm not too sure about how `minimumShouldMatch` works in this case, we will have to check.

Another this that struck me just now when thinking about it: Not sure if we also need to handle this case in the  `toQuery()` itself, because query doesn't need to be parsed always in the future. The problem here is that at least for now we are using the `isFilter()` state in the QueryParseContex when using it in `toQuery()`, so there currently we don't know if the current query is used as a filter or not in the parent query. Both those issues don't have anything to do with introducing EmptyQueryBuilder here but with the fact that the decison to set `minimumShouldMatch` if not set explicitely was based on the presence of an actual lucene query so far, which really we can only know in the query-generation phase (aka toQuery). 
Since the whole conditional was moved over from the former BoolQueryFilter, it might be a good thing to ask if there is another way to solve this and why doing this only in filter context is necessary. If we could do it also for non-filter context it would be easy to move this to `toQuery`